### PR TITLE
[CARBONDATA-726] Handled query and scan for V3 format.

### PR DESCRIPTION
**Pull request by ravipesala**
1. Reader interfaces DimensionColumnChunkReader and MeasureColumnChunkReader are updated to separate IO and processing.
2. Updated Reader interfaces to handle pages inside data blocks.
3.  Separated out threads for read and write.
4. Added read a head blocklet.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/811/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/816/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/817/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/818/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/819/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/820/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/821/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/827/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/829/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/932/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/933/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/934/

File: core/src/main/java/org/apache/carbondata/core/datastore/impl/btree/BTreeNonLeafNode.java [hunk 1]
```diff
@@ -24,8 +24,8 @@
 import org.apache.carbondata.core.datastore.DataRefNode;
 import org.apache.carbondata.core.datastore.FileHolder;
 import org.apache.carbondata.core.datastore.IndexKey;
-import org.apache.carbondata.core.datastore.chunk.DimensionColumnDataChunk;
-import org.apache.carbondata.core.datastore.chunk.MeasureColumnDataChunk;
+import org.apache.carbondata.core.datastore.chunk.impl.DimensionRawColumnChunk;
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/datastore/impl/btree/BTreeNonLeafNode.java]: Why putting `DimensionRawColumnChunk` under `impl` package? It is exposed like interface right?

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RowLevelFilterExecuterImpl.java [hunk 2]
```diff
@@ -343,24 +339,39 @@ private int readSurrogatesFromColumnGroupBlock(BlocksChunkHolder blockChunkHolde
...
+              blockChunkHolder.getDataBlock()
+                  .getDimensionChunk(blockChunkHolder.getFileReader(), blocksIndex[i]);
+        }
+      } else {
+        GenericQueryType complexType = complexDimensionInfoMap.get(blocksIndex[i]);
+        complexType.fillRequiredBlockData(blockChunkHolder);
+      }
+    }
+
+    // CHECKSTYLE:OFF Approval No:Approval-V1R2C10_001
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RowLevelFilterExecuterImpl.java]: remove this

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RestructureFilterExecuterImpl.java [hunk 3]
```diff
@@ -35,18 +37,18 @@ public RestructureFilterExecuterImpl(DimColumnResolvedFilterInfo dimColumnResolv
...
   }
 
   @Override public BitSet isScanRequired(byte[][] blockMaxValue, byte[][] blockMinValue) {
     BitSet bitSet = new BitSet(1);
     bitSet.set(0);
     return bitSet;
   }
+
+  @Override public void readBlocks(BlocksChunkHolder blockChunkHolder) throws IOException {
+    // TODO
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RestructureFilterExecuterImpl.java]: Throw UnSupportedOperationException instead of this?

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RowLevelRangeLessThanEqualFilterExecuterImpl.java [hunk 4]
```diff
@@ -53,11 +55,19 @@ public RowLevelRangeLessThanEqualFilterExecuterImpl(
     BitSet bitSet = new BitSet(1);
     byte[][] filterValues = this.filterRangeValues;
     int columnIndex = this.dimColEvaluatorInfoList.get(0).getColumnIndex();
+    boolean isScanRequired = isScanRequired(blockMinValue[columnIndex], filterValues);
+    if (isScanRequired) {
+      bitSet.set(0);
+    }
+    return bitSet;
+
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RowLevelRangeLessThanEqualFilterExecuterImpl.java]: remove empty line

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RowLevelRangeLessThanFiterExecuterImpl.java [hunk 5]
```diff
@@ -53,11 +55,19 @@ public RowLevelRangeLessThanFiterExecuterImpl(
     BitSet bitSet = new BitSet(1);
     byte[][] filterValues = this.filterRangeValues;
     int columnIndex = this.dimColEvaluatorInfoList.get(0).getColumnIndex();
+    boolean isScanRequired = isScanRequired(blockMinValue[columnIndex], filterValues);
+    if (isScanRequired) {
+      bitSet.set(0);
+    }
+    return bitSet;
+
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RowLevelRangeLessThanFiterExecuterImpl.java]: remove empty line

File: core/src/main/java/org/apache/carbondata/core/scan/result/iterator/AbstractDetailQueryResultIterator.java [hunk 6]
```diff
@@ -191,8 +192,13 @@ public void processNextBatch(CarbonColumnarBatch columnarBatch) {
...
 
   @Override public void close() {
-    CarbonUtil.freeMemory(blocksChunkHolder.getDimensionDataChunk(),
-        blocksChunkHolder.getMeasureDataChunk());
+    try {
+      fileReader.finish();
+    } catch (IOException e) {
+      LOGGER.error(e);
+    }
+    CarbonUtil.freeMemory(blocksChunkHolder.getDimensionRawDataChunk(),
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/scan/result/iterator/AbstractDetailQueryResultIterator.java]: put it in finally block

File: core/src/main/java/org/apache/carbondata/core/util/BitSetGroup.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.carbondata.core.util;
+
+import java.util.BitSet;
+
+/**
+ * Maintains the group of bitsets.
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/util/BitSetGroup.java]: Can you add something like "It is used for .... while decoding data chunk"

File: core/src/main/java/org/apache/carbondata/core/cache/dictionary/Dictionary.java [hunk 8]
```diff
@@ -59,6 +59,17 @@
   String getDictionaryValueForKey(int surrogateKey);
 
   /**
+   * This method will find and return the dictionary value for a given surrogate key in bytes.
+   * Applicable scenarios:
+   * 1. Query final result preparation : While convert the final result which will
+   * be surrogate key back to original dictionary values this method will be used
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/cache/dictionary/Dictionary.java]: It is added because `getDictionaryValueForKey` gets the bytes from dictionary and convert to String, but in our decoder we again convert to bytes. So I added new method to return bytes directly with out converting to String to avoid multiple conversions.

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java [hunk 9]
```diff
@@ -589,7 +591,7 @@
    * INMEMORY_REOCRD_SIZE
    */
   public static final String DETAIL_QUERY_BATCH_SIZE = "carbon.detail.batch.size";
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java]: ok

File: core/src/main/java/org/apache/carbondata/core/datastore/FileHolder.java [hunk 10]
```diff
@@ -18,8 +18,12 @@
 package org.apache.carbondata.core.datastore;
 
 import java.io.IOException;
+import java.nio.ByteBuffer;
 
 public interface FileHolder {
+
+  void readByteBuffer(String filePath, ByteBuffer byteBuffer, long offset, int length)
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/datastore/FileHolder.java]: ok

File: core/src/main/java/org/apache/carbondata/core/datastore/chunk/AbstractRawColumnChunk.java [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.carbondata.core.datastore.chunk;
+
+import java.nio.ByteBuffer;
+
+
+/**
+ * It contains group of uncompressed blocklets on one column.
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/datastore/chunk/AbstractRawColumnChunk.java]: The main benefit of V3 is to avoid multiple IO requests, in V3 format each column chunk has group of column chunks of size 32000 called as pages. This group( group size can be configured ) is stored continuously so for one IO request it gets complete column chunk with all pages.

Yes this description will be added in V3 reader/ writer. That will be another PR added by Vishal.

Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/datastore/chunk/AbstractRawColumnChunk.java]: And for V1 and V2 formats always has one page for backward compatibility.

File: core/src/main/java/org/apache/carbondata/core/datastore/chunk/reader/DimensionColumnChunkReader.java [hunk 12]
```diff
@@ -35,7 +36,7 @@
    * @param blockIndexes blocks to be read
    * @return dimension column chunks
    */
-  DimensionColumnDataChunk[] readDimensionChunks(FileHolder fileReader, int[][] blockIndexes)
+  DimensionRawColumnChunk[] readRawDimensionChunks(FileHolder fileReader, int[][] blockIndexes)
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/datastore/chunk/reader/DimensionColumnChunkReader.java]: Yes, it is bloclklet index not blockindex. I will update the same

File: core/src/main/java/org/apache/carbondata/core/datastore/chunk/reader/dimension/v1/CompressedDimensionChunkFileBasedReaderV1.java [hunk 13]
```diff
@@ -56,51 +58,74 @@ public CompressedDimensionChunkFileBasedReaderV1(final BlockletInfo blockletInfo
...
+    byte[] data = new byte[dimensionRawColumnChunk.getLength()];
+    rawData.get(data);
+    dataPage = COMPRESSOR.unCompressByte(data);
 
-    // first read the data and uncompressed it
-    dataPage = COMPRESSOR.unCompressByte(fileReader
-        .readByteArray(filePath, dimensionColumnChunk.get(blockIndex).getDataPageOffset(),
-            dimensionColumnChunk.get(blockIndex).getDataPageLength()));
     // if row id block is present then read the row id chunk and uncompress it
     if (CarbonUtil.hasEncoding(dimensionColumnChunk.get(blockIndex).getEncodingList(),
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/datastore/chunk/reader/dimension/v1/CompressedDimensionChunkFileBasedReaderV1.java]: ok

File: core/src/main/java/org/apache/carbondata/core/datastore/chunk/reader/dimension/v1/CompressedDimensionChunkFileBasedReaderV1.java [hunk 14]
```diff
@@ -56,51 +58,74 @@ public CompressedDimensionChunkFileBasedReaderV1(final BlockletInfo blockletInfo
...
         Encoding.INVERTED_INDEX)) {
+      byte[] columnIndexData;
+      synchronized (fileReader) {
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/datastore/chunk/reader/dimension/v1/CompressedDimensionChunkFileBasedReaderV1.java]: it is required because IO is handled in different thread and blocklet processing handled in different thread. But in some cases like after filter processing the required projection blockets can be read in processing thread also. So that is why this synchronization is required.

File: core/src/main/java/org/apache/carbondata/core/datastore/chunk/reader/dimension/v2/CompressedDimensionChunkFileBasedReaderV2.java [hunk 15]
```diff
@@ -118,54 +122,109 @@ public CompressedDimensionChunkFileBasedReaderV2(final BlockletInfo blockletInfo
...
     if (hasEncoding(dimensionColumnChunk.encoders, Encoding.RLE)) {
+      byte[] dataRle = new byte[dimensionColumnChunk.rle_page_length];
+      rawData.position(copySourcePoint);
+      rawData.get(dataRle);
       rlePage =
-          numberComressor.unCompress(data, copySourcePoint, dimensionColumnChunk.rle_page_length);
+          numberComressor.unCompress(dataRle, 0, dimensionColumnChunk.rle_page_length);
       // uncompress the data with rle indexes
       dataPage = UnBlockIndexer.uncompressData(dataPage, rlePage, eachColumnValueSize[blockIndex]);
       rlePage = null;
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/datastore/chunk/reader/dimension/v2/CompressedDimensionChunkFileBasedReaderV2.java]: It is from old code, it is not rquired and removed now

File: core/src/main/java/org/apache/carbondata/core/datastore/chunk/reader/measure/v1/CompressedMeasureChunkFileBasedReaderV1.java [hunk 16]
```diff
@@ -77,20 +79,38 @@ public CompressedMeasureChunkFileBasedReaderV1(final BlockletInfo blockletInfo,
    * @param blockIndex block to be read
    * @return measure data chunk
    */
-  @Override public MeasureColumnDataChunk readMeasureChunk(final FileHolder fileReader,
-      final int blockIndex) throws IOException {
+  @Override public MeasureRawColumnChunk readRawMeasureChunk(FileHolder fileReader, int blockIndex)
+      throws IOException {
+    ByteBuffer buffer =
+        ByteBuffer.allocateDirect(measureColumnChunks.get(blockIndex).getDataPageLength());
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/datastore/chunk/reader/measure/v1/CompressedMeasureChunkFileBasedReaderV1.java]: ok

File: core/src/main/java/org/apache/carbondata/core/datastore/impl/btree/BTreeNonLeafNode.java (hunk 1 above)
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/datastore/impl/btree/BTreeNonLeafNode.java]: `DimensionRawColumnChunk` is not interface, it is implemented class

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RowLevelFilterExecuterImpl.java (hunk 2 above)
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RowLevelFilterExecuterImpl.java]: ok

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RestructureFilterExecuterImpl.java (hunk 3 above)
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RestructureFilterExecuterImpl.java]: This class is not required, removed it

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RowLevelRangeLessThanEqualFilterExecuterImpl.java (hunk 4 above)
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RowLevelRangeLessThanEqualFilterExecuterImpl.java]: ok

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RowLevelRangeLessThanFiterExecuterImpl.java (hunk 5 above)
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RowLevelRangeLessThanFiterExecuterImpl.java]: ok

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/937/

**Comment by jackylk** [comment]
LGTM
